fix: fixed bad text slicing for Paragraph's lines

chopan050 opened the pull request:
<!-- Thank you for contributing to Manim! Learn more about the process in our contributing guidelines: https://docs.manim.community/en/latest/contributing.html -->

## Overview: What does this pull request change?
<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->
<!--changelog-start-->
Inside `Paragraph`'s `_gen_chars` method, a bad string slicing caused the text's characters to be obtained in an incorrect order. This would cause `chars` (and thus `self.chars` and `self.lines` in `__init__`) to store incorrect values.
If `Paragraph`'s `alignment` parameter was None, nothing wrong seemed to happen, but if `alignment` was 'left', 'center' or 'right', the text's lines would not display correctly.
This happened because the slicing was of the form `[begin : begin + string_length + 1]`. This bug was fixed by changing it to `[begin : begin + char_count]`, where `char_count` is the number of characters in the corresponding string which are not " ", "\n" or "\t".

| Before | After |
| ------- | ------ |
| ![Paragraph Before](https://user-images.githubusercontent.com/49853152/168176717-f8ff483b-2f17-4875-8e59-5ff88c6f785a.png)| ![Paragraph After](https://user-images.githubusercontent.com/49853152/168176710-a723912b-5e2b-4e8d-ab54-1fd48dca43bb.png) |

<details><summary>Click here to see the test scene code:</summary>

```python
class ParagraphScene(Scene):
    def construct(self):
        pars = []
        for alignment in (None, "left", "center", "right"):
            pars.append(VGroup(
                Text("alignment: {}".format(alignment), color=YELLOW).scale(1.25),
                Paragraph("line_with_no_spaces", "1234567890", alignment=alignment),
                Paragraph("line with three spaces", "1234567890", alignment=alignment),
            ).arrange_submobjects(DOWN, buff=0.5).scale(0.5))

        pars[0].move_to(3*LEFT + 2*UP)
        pars[1].move_to(3*RIGHT + 2*UP)
        pars[2].move_to(3*LEFT + 2*DOWN)
        pars[3].move_to(3*RIGHT + 2*DOWN)

        self.add(*pars)
```
</details>

I've also cleaned `__init__`'s code:
- There was a redundant fragment of code which did exactly the same thing as `_gen_chars`, storing data in `chars_lines_text_list`. I deleted this code and used `self.chars` instead.
- `lines_str` joins the text strings with '\n', but then `lines_str_list` split `lines_str` again, with '\n' as delimiter. I rewrote `lines_str_list` as just `list(text)`.
- `self.lines` is a list which essentially stores `list(self.chars)` as its first element. The original code used a for loop to append `self.chars`' lines, one by one, to `self.lines[0]`. I replaced this for loop with `list(self.chars)`.
- `self.lines`' second element was rewritten as a repetition of `[self.alignment]`, `self.chars.__len()__` times.
- I replaced the for loop used to append the line centers to `self.lines_initial_positions`, and instead used `[line.get_center() for line in self.lines[0]]`.
<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->

## Links to added or changed documentation pages
<!-- Please add links to the affected documentation pages (edit the description after opening the PR). The link to the documentation for your PR is https://manimce--####.org.readthedocs.build/en/####/, where #### represents the PR number. -->
https://manimce--2721.org.readthedocs.build/en/2721/


## Further Information and Comments
<!-- If applicable, put further comments for the reviewers here. -->



<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

behackl reviewed (approved):
This looks like a very nice improvement to me, thank you very much for your efforts!

I can remember that I got the alignment parameter to work correctly in the past by passing `disable_ligatures=True` -- could you please check whether this is the case here as well? (Edit: that the alignment still works with disabled ligatures, that is.)

Theoretically, it should also be possible to write a non-graphical test for the alignment argument; perhaps something along the lines of

```
def test_paragraph_alignment():
    par_left = Paragraph("This is", "a left-aligned", "paragraph", alignment="left")
    assert par_left[0][0].get_x(LEFT) == par_left[1][0].get_x(LEFT) == par_left[2][0].get_x(LEFT)

    par_right = Paragraph("This is", "a right-aligned", "paragraph", alignment="right")
    assert par_right[0][-1].get_x(RIGHT) == par_right[1][-1].get_x(RIGHT) == par_right[2][-1].get_x(RIGHT)
```

What do you think?

chopan050 commented:
I found some interesting things, and tried some changes. Let me know what you think about them:

As `Paragraph` is a subclass of `Text`, I need to talk about the latter. When `Text.disable_ligatures` is False, then " ", "\n" and "\t" are not counted as "characters" and thus are not stored in `Text.chars` nor `Text.submobjects`. This was evidenced in my original post, where I had to remove these blank spaces from my `char_count` variable to index my `Paragraph.lines_text.chars` correctly.

However, when `Text.disable_ligatures` is True, then it does count them as characters in `Text.chars` and `Text.submobjects`. Inside `Text.__init__`:

```python
super().__init__(...)
self.text = text
if self.disable_ligatures:
    self.submobjects = [*self._gen_chars()]
self.chars = self.get_group_class()(*self.submobjects)
```

This calls `super().__init__`, which among other things initializes `self.submobjects`. But then it asks whether `self.disable_ligatures` is True or False. If it's True, then it calls `self._gen_chars`, assigns its value to `self.submobjects` as a list, and then assigns the same value to `self.chars` as a group. The thing is, `Text._gen_chars` actually considers spaces as chars:

```python
def _gen_chars(self):
    chars = self.get_group_class()()
    submobjects_char_index = 0
    for char_index in range(self.text.__len__()):
        if self.text[char_index] in (" ", "\t", "\n"):
            space = Dot(radius=0, fill_opacity=0, stroke_opacity=0)
            # some code for positioning this space...
            chars.add(space) # <--- It's considered as a char!
        else:
            chars.add(self.submobjects[submobjects_char_index])
            submobjects_char_index += 1
    return chars
```

So, first of all: is this behavior a bug? Is it intended that when `Text.disabled_ligatures` is True then spaces are considered in `Text.chars`, but when it's False then they are not?

Here's what I did:

- On one hand, I assumed it's a bug, and I assumed that spaces should never be counted as chars (this `submobjects_char_index += 1` line, which is present only inside the else statement, when the character is not a space, made me think so). Thus, I removed the `chars.add(space)` line, and tested my original code again in both cases, whether `disabled_ligatures` is True or False. It worked perfectly on both scenarios! :D

- On the other hand, if this is not a bug and it's actually intended, then what I have to do is to decide whether to count spaces or not, depending on whether disable_ligatures is True or False. However, `Paragraph._gen_chars` currently does not have a way of knowing about that information. So I added another attribute, `Paragraph.consider_spaces_as_chars = config["disable_ligatures"]`, and modified `Paragraph._gen_chars`' method so that it removes spaces from the count only if `Paragraph.consider_spaces_as_chars` is False. However, this solution didn't go well: the lines are strangely aligned when `disabled_ligatures` is True. D:

The following scenarios are for `disabled_ligatures == True`. In both cases, `disabled_ligatures == False` still works fine.

| by deleting `chars.add(space)` from `Text._gen_chars` | by adding `Paragraph.consider_spaces_as_chars` |
| ------- | ------ |
| ![imagen](https://user-images.githubusercontent.com/49853152/169634553-fbdbf757-96a5-4f06-a203-0de9326299cc.png) |![imagen](https://user-images.githubusercontent.com/49853152/169634603-819efcba-f896-403e-b248-d53dec8aba0b.png) |
| Works good! | You can notice that when alignment == "center" some lines are slightly shifted to the right and when alignment == "right" these lines are a bit more shifted to the right, but that's it. I haven't figured out what's going on.

Well, what do you think about it? Is it a bug or is it intended?

behackl commented:
> So, first of all: is this behavior a bug? Is it intended that when `Text.disabled_ligatures` is True then spaces are considered in `Text.chars`, but when it's False then they are not?

This is not entirely obvious to answer, unfortunately (mainly because the implementation is so messy and details and intentions are not really documented). The advantage of having a mode where spaces are counted as additional characters is that applying a slice to the generated mobject then has the same effect as applying it to the original string (which is useful for coloring parts of your rendered string, for example). Personally, I disagree with hiding this behavior behind `disable_ligatures` (because, as far as I know, this feature of preserving slices is not documented explicitly anywhere, and semantically it is also not clear at all).

At this point it does not make sense to make far-reaching changes to `Text` (because the interface to ManimPango will is being reworked, cc @naveen521kk), but it would still be nice to come to a better solution here somehow.

Do you think some sort of compromise would be possible? For example, by creating a new keyword argument (`consider_spaces_as_chars`?) for which this behavior of preserving slices could then be documented explicitly (along the lines of "if you want to access parts of the string in the same way as the original string, set both `disable_ligatures=True` and `consider_spaces_as_chars=True` -- in this constellation, the alignment argument might not work correctly"). I realize that this does not necessarily make the implementation of `Text` any easier, but it might be a suitable solution until the text API is reworked overall.

What do you think about this?

chopan050 commented:
> This is not entirely obvious to answer, unfortunately (mainly because the implementation is so messy and details and intentions are not really documented). The advantage of having a mode where spaces are counted as additional characters is that applying a slice to the generated mobject then has the same effect as applying it to the original string (which is useful for coloring parts of your rendered string, for example). Personally, I disagree with hiding this behavior behind disable_ligatures (because, as far as I know, this feature of preserving slices is not documented explicitly anywhere, and semantically it is also not clear at all).

I see, so it makes sense to have the choice to consider spaces as chars.

>Do you think some sort of compromise would be possible? For example, by creating a new keyword argument (consider_spaces_as_chars?) for which this behavior of preserving slices could then be documented explicitly (along the lines of "if you want to access parts of the string in the same way as the original string, set both disable_ligatures=True and consider_spaces_as_chars=True -- in this constellation, the alignment argument might not work correctly"). I realize that this does not necessarily make the implementation of Text any easier, but it might be a suitable solution until the text API is reworked overall.

I was wondering, how would this new keyword argument `consider_spaces_as_chars` work exactly? I mean, I could do the same thing I did before (to decide whether to count spaces or not in `Paragraph._gen_chars`' `char_count` variable). But with this new piece of information ("strings can be indexed and sliced to apply different properties to these slices"), I was wondering if this new kwarg should actually modify something else. Like, if I add it to `Text`, should I do something like this inside `Text._gen_chars`?

```python
if self.consider_spaces_as_chars:
    chars.add(space)
```

If I do this, then I probably should also modify `SVGMobject`'s implementation (I think), because currently (when `disable_ligatures` is False) it never considers spaces as actual chars or submobjects. `SVGMobject`'s inner workings seem more complex than what I can handle right now lol.

If I do nothing else, then this new `consider_spaces_as_chars` kwarg would be kind of broken, because right now it's `disabled_ligatures` the argument which directly controls whether spaces are considered as chars or not, and that's pretty complex as well.

I don't really know, this whole thing ended up being a lot more complex than I thought lol

What do you think about this?

naveen521kk commented:
> At this point it does not make sense to make far-reaching changes to Text (because the interface to ManimPango will is being reworked, cc @naveen521kk), but it would still be nice to come to a better solution here somehow.

Waiting for the new Text API is better as all this will be handled by the underlying library directly. How long it'll take to be done depends on how much time I've for manim; which will be less this month. There's a milestone which tracks them https://github.com/ManimCommunity/ManimPango/milestone/1 (the core rendering into an SVG part is yet to be done, while others are done).

When `disable_ligatures` is disabled we can't iterate the text and hope it'll be equal to the number of characters actually rendered from the SVG, so we don't iterate it at all (that is `Text._gen_chars` is not called at all). When `disable_ligatures` is True most likely the number of characters is the same as the actual text at least for English characters, we count the spaces and add them to the submobjects list. So when `disable_ligatures` is True one can iterate the submobject and expect it to be equal to the text passed. I remember documenting this behaviour somewhere but can't find it. Hope this clears the confusion.

chopan050 commented:
>naveen521kk: When disable_ligatures is disabled we can't iterate the text and hope it'll be equal to the number of characters actually rendered from the SVG, so we don't iterate it at all (that is Text._gen_chars is not called at all). When disable_ligatures is True most likely the number of characters is the same as the actual text at least for English characters, we count the spaces and add them to the submobjects list. So when disable_ligatures is True one can iterate the submobject and expect it to be equal to the text passed. I remember documenting this behaviour somewhere but can't find it. Hope this clears the confusion.

I see. So, from what I understood:
- Rendering only creates non-blank characters. Spaces are not rendered, and `Text._gen_chars` circumvents that by adding invisible `Dot`s to the list of submobjects, right?
- When `disable_ligatures` is True, we do expect every submobject to correspond to every character, including spaces. Thus spaces should be counted, and `chars.add(space)` should stay (where `space` is an invisible `Dot`).
- When `disable_ligatures` is False, we don't really expect that. The text should not be iterated over.

>Benjamin Hackl: but it would still be nice to come to a better solution here somehow.
>naveen521kk: Waiting for the new Text API is better as all this will be handled by the underlying library directly. 

On one hand, it's true that waiting for this new `Text` rework is better to fix these issues. On the other hand, I still think that making this small change for `Paragraph` is good, at least for the moment.

So, I'll continue with the idea of creating `Paragraph.consider_spaces_as chars = config.get("disable_ligatures", False)` and deciding whether to remove spaces or not depending on its value. Some things I realized about this:

- If all spaces are considered as chars, including line breaks (\n), then I have to increment `char_index_counter` by 1 in every single iteration. That's because there's a line break character separating every line in the text, which I was not counting before.
- It turns out that this "weird alignment" that happens when `disable_ligatures` is True, occurs when there are leading spaces in the lines we need to align. That's why it always happened before: because I didn't increment `char_index_counter` by 1, the first character in the second line was always a \n, which triggered this "weird alignment". Now that I fixed this `char_index_counter` issue, let's compare what happens in these two scenarios, where the second scenario has a leading space in the second line:

|`Paragraph("Hello there", "a", alignment="center", disable_ligatures=True)` | `Paragraph("Hello there", " a", alignment="center", disable_ligatures=True)` |
| --- | --- |
| ![imagen](https://user-images.githubusercontent.com/49853152/169709161-777f02ec-840c-4808-8ff3-7fe03e20cbf5.png) | ![imagen](https://user-images.githubusercontent.com/49853152/169709210-061ccf37-1ced-428f-821a-9c4202f70fd9.png) |

(this only happens when `disable_ligatures` is True. Otherwise, the leading spaces are simply ignored and alignment works as expected)

At this point tho, I don't really think it's worth it to fix this behavior when there are leading spaces. After all, `Text` is going to be reworked anyways. But I think that the changes I already made are worth it, at least for now. So I'll commit the changes to my branch.

Final note: I also reversed one of the first changes I made:

>`lines_str` joins the text strings with '\n', but then `lines_str_list` split `lines_str` again, with '\n' as delimiter. I rewrote `lines_str_list` as just `list(text)`.

Actually, if I put a \n in one of the strings I pass to `Paragraph` as arguments, it was actually convenient to join the strings and then split them again. So I reversed this change I made.

What do you think?

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -147,47 +147,29 @@ class Paragraph(VGroup):
     def __init__(self, *text, line_spacing=-1, alignment=None, **config):
```
I don't like that the additional keyword arguments are named `config` here, I'd rather see them be named `kwargs` as everywhere else. This is more venting than anything else though, not a change request. :-)

chopan050 replied:
Huh, for a moment I thought that `config` was used more often. I now checked and you're right, pretty much everywhere else they're named `kwargs`. To be more consistent, yeah, I would change it to `kwargs`.

chopan050 replied:
I just noticed that using `config` as a name shadows the `config` global variable, so this is actually an issue which should be solved. I'd definitely change it to `kwargs`.

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -147,47 +147,29 @@ class Paragraph(VGroup):
-        self.lines[1].extend(
-            [self.alignment for _ in range(chars_lines_text_list.__len__())],
-        )
+        self.lines = [list(self.chars), [self.alignment] * self.chars.__len__()]
```
I don't think it is particularly pythonic to get the length by directly calling the dunder method. This is more readable.

```suggestion
        self.lines = [list(self.chars), [self.alignment] * len(self.chars)]
```

chopan050 replied:
The original source code used this `__len__()` method. I've always found it really weird, but didn't change it because of... reasons? I agree with you tho, using `len(self.chars)` is indeed more readable.

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -147,47 +147,29 @@ class Paragraph(VGroup):
 
     def _gen_chars(self, lines_str_list):
-        """Function to convert plain string to 2d-VGroup of chars. 2d-VGroup mean "VGroup of VGroup".
+        """Function to convert a list of plain strings to a 2d-VGroup of chars.
+        "2d-VGroup" means "VGroup of VGroups".
```
I think the wording here is worth improving.

```suggestion
        """Function to convert a list of plain strings to a VGroup containing VGroups.
```

chopan050 replied:
True, I wasn't really convinced by the use of "2d-VGroup" in this docstring. Again, I didn't change it so much when I submitted this PR, but again I agree with you on this one.
Maybe I'd add "of chars" at the end of your suggestion, something like
"Function to convert a list of plain strings to a VGroup containing VGroups of chars."

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -147,47 +147,29 @@ class Paragraph(VGroup):
         if self.alignment:
             self._set_all_lines_alignments(self.alignment)
 
     def _gen_chars(self, lines_str_list):
```
Type hints should go in method header instead of docstring

```suggestion
    def _gen_chars(self, lines_str_list: list) -> VGroup:
```

chopan050 replied:
You're right, I'll change it then.

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -147,47 +147,29 @@ class Paragraph(VGroup):
         ----------
-        lines_str_list : :class:`str`
-            Plain text string.
+        lines_str_list : :class:`list`
```
```suggestion
        lines_str_list
```

chopan050 replied:
Same as before

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -197,15 +179,24 @@ def _gen_chars(self, lines_str_list):
         char_index_counter = 0
         chars = self.get_group_class()()
         for line_no in range(lines_str_list.__len__()):
```
While we are at it... 😅 

```suggestion
        for line_no in range(len(lines_str_list)):
```

chopan050 replied:
😅

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -197,15 +179,24 @@ def _gen_chars(self, lines_str_list):
+            line_str = lines_str_list[line_no]
+            # Count all the characters in line_str
+            # Spaces may or may not count as characters
+            char_count = line_str.__len__()
```
```suggestion
            char_count = len(line_str)
```

chopan050 replied:
Yup, same as before

behackl commented on manim/mobject/text/text_mobject.py:
```diff
@@ -197,15 +179,24 @@ def _gen_chars(self, lines_str_list):
+            # Spaces may or may not count as characters
+            char_count = line_str.__len__()
+            if not self.consider_spaces_as_chars:
+                for blank in (" ", "\n", "\t"):
+                    char_count -= line_str.count(blank)
```
Instead of hardcoding these particular spaces, why not use the inbuilt `isspace` method of strings? It also detects `\f` and `\r`, which should make this even more robust.

```suggestion
                for char in line_str:
                    if char.isspace():
                        char_count -= 1
```

behackl replied:
In case you prefer a more concise version:

```suggestion
                char_count -= len([char for char in line_str if char.isspace()])
```

chopan050 replied:
Ooh, I didn't know there was an `isspace` method for strings! That's new for me 😮 
(also I didn't know about `\f` and `\r`)
I like the first suggestion, because it's more readable.
However, before I apply it, should I change `Text._gen_chars` as well? Because:

```python
    def _gen_chars(self):
        chars = self.get_group_class()()
        submobjects_char_index = 0
        for char_index in range(self.text.__len__()):
            if self.text[char_index] in (" ", "\t", "\n"):
                space = Dot(radius=0, fill_opacity=0, stroke_opacity=0)
                if char_index == 0:
                    space.move_to(self.submobjects[submobjects_char_index].get_center())
                else:
                    space.move_to(
                        self.submobjects[submobjects_char_index - 1].get_center(),
                    )
                chars.add(space)
            else:
                chars.add(self.submobjects[submobjects_char_index])
                submobjects_char_index += 1
        return chars
```

Besides this `self.text.__len__()` being used in the for loop (and some others in `Paragraph` lol), there's this condition `if self.text[char_index] in (" ", "\t", "\n"):`. I should probably also change this to `if self.text[char_index].isspace():`, right?

behackl reviewed (changes requested):
Sorry, took me a while to get back to you again -- I agree, I think the order of magnitude of changes here is worth getting merged before the text API change.

I've left a bunch of minor suggestions, please take a look and let me know what you think; afterwards I'd be happy with merging this.

chopan050 commented:
I've addressed your suggestions, what do you think of the changes? Let me know.

MrDiver reviewed (approved):
approving in the name of @behackl due to previous comment

behackl reviewed (approved):
Sorry for the long wait, I think this is fine! Thank you very much for your contribution!

MrDiver commented:
Btw can we add tests to this that we do not further break this functionality in the future ?

I would probably suggest opening a new Issue for that!

MrDiver commented:
- #2839

I have created a new issue for addressing the tests. I think this functionality can now be merged!
